crates/tx-graph2/src/connectors/claim_contest_connector.rs

```diff
@@ -0,0 +1,439 @@
+    }
+
+    /// Generates a vector of all leaf scripts of the connector.
+    pub fn leaf_scripts(&self) -> Vec<ScriptBuf> {
+        let mut scripts = Vec::new();
+
+        for watchtower_pubkey in &self.watchtower_pubkeys {
+            let script = bitcoin::script::Builder::new()
+                .push_slice(self.n_of_n_pubkey.serialize())
+                .push_opcode(opcodes::all::OP_CHECKSIGVERIFY)
+                .push_slice(watchtower_pubkey.serialize())
+                .push_opcode(opcodes::all::OP_CHECKSIG)
+                .into_script();
+            scripts.push(script);
+        }
+
+        let script = bitcoin::script::Builder::new()
+            .push_slice(self.n_of_n_pubkey.serialize())
+            .push_opcode(opcodes::all::OP_CHECKSIGVERIFY)
+            .push_sequence(self.delta_contest.to_sequence())
+            .push_opcode(opcodes::all::OP_CSV)
+            .into_script();
+        scripts.push(script);
+
+        scripts
+    }
+
+    /// Generates the bitcoin address and taproot spending info of the connector.
+    pub fn address_and_spend_info(&self) -> (bitcoin::Address, TaprootSpendInfo) {
+        let internal_key = self.n_of_n_pubkey;
+        let scripts = self.leaf_scripts();
+        create_taproot_addr(
+            &self.network,
+            SpendPath::Both {
+                internal_key,
+                scripts: scripts.as_slice(),
+            },
+        )
+        .expect("tap tree is valid")
+    }
+
+    /// Generates the txout of the connector.
+    pub fn tx_out(&self) -> TxOut {
+        let script_pubkey = self.address_and_spend_info().0.script_pubkey();
+        let dust = script_pubkey.minimal_non_dust();
+        // TODO: Replace magic number 3 with constant from contest transaction, once the code exists
+        let value = dust * (3 * self.n_watchtowers() as u64);
+
+        TxOut {
+            value,
+            script_pubkey,
+        }
+    }
+
+    /// Finalizes the PSBT `input` where the connector is used, using the provided `witness`.
+    ///
+    /// # Warning
+    ///
+    /// If [`ClaimContestSpendPath::Uncontested`] is used, then the sequence of the transaction
+    /// input must be set accordingly. Also, the global locktime has to be set accordingly.
+    pub fn finalize_input(&self, input: &mut Input, witness: ClaimContestWitness) {
+        let ClaimContestWitness {
+            n_of_n_signature,
+            spend_path,
+        } = witness;
+        let taproot_spend_info = self.address_and_spend_info().1;
+        let leaf_index = match spend_path {
+            ClaimContestSpendPath::Contested {
+                watchtower_index,
+                watchtower_signature: _,
+            } => watchtower_index as usize,
+            ClaimContestSpendPath::Uncontested => self.n_watchtowers(),
+        };
+        let leaf_script = self
+            .leaf_scripts()
+            .into_iter()
+            .nth(leaf_index)
+            .expect("leaf script exists");
```

Comment:
If you are just panicking, you can just index directly.

```suggestion
        let leaf_script = self.leaf_scripts()[leaf_index];
```

I don't think the panic message is any worse even here. In fact, it's more useful to us if rust tells us exactly which index is missing instead of obscuring it with the expect message.